### ISSUE TITLE
Lisp URL related issues

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#2027]
I've noticed a few issues with Lisp URLs and  since the fix is probably common to all of them, let me list them here:

1. Looks like `lisp-url` should no longer be used.  Its only call site is `send-lisp-url`.  Move it there?

2. Actually `download.lisp` still uses `lisp-url` and as such all buttons are broken.

3. If I got it right, this means that `user-interface:button` can be removed.

4. The following snippet is faulty in `gtk.lisp`:

```lisp
;; Objects and other complex structures make cl-json choke.
(unless (or (typep result 'standard-object)
                                       (and (typep result 'cons)
                                            (some #'listp result)))
                             (cl-json:encode-json-to-string result))
```

It does not take provisions for non-cons sequences, nor for standard-objects nor structs in these sequences.  Something as simple as `(list (quri:quri "atlas.engineer"))` fails here.

Proposed solution:

```lisp
;; `sb-pcl:slot-object` has these 3 subclasses.
(defun slot-object-p (result)  (typep result '(or standard-object structure-object condition)))

; ...

(unless (or (slot-object-p result)
                                       (and (typep result 'sequence)
                                            (some #'slot-object-p  result)))
                             (cl-json:encode-json-to-string result))
```

@aartaka What do you think?

EDIT: Typos.

**Comment by Ambrevar** [comment]
I also propose moving the condition of the `unless` chunk to a `compound-object-p` function in the line of `sb-int:compount-object-p`.

**Comment by aartaka** [comment]
> 4\. The following snippet is faulty in `gtk.lisp`:
> 
> ```lisp
> ;; Objects and other complex structures make cl-json choke.
> (unless (or (typep result 'standard-object)
>                                        (and (typep result 'cons)
>                                             (some #'listp result)))
>                              (cl-json:encode-json-to-string result))
> ```
> 
> It does not take provisions for non-cons sequences, nor for standard-objects nor structs in these sequences. Something as simple as `(list (quri:quri "atlas.engineer"))` fails here.
> 
> Proposed solution:
> 
> ```lisp
> ;; `sb-pcl:slot-object` has these 3 subclasses.
> (defun slot-object-p (result)  (typep result '(or standard-object structure-object condition)))
> 
> ; ...
> 
> (unless (or (slot-object-p result)
>                                        (and (typep result 'sequence)
>                                             (some #'slot-object-p  result)))
>                              (cl-json:encode-json-to-string result))
> ```
> 
> @aartaka What do you think?

Hash-tables are to be checked too. Actually, is there an umbrella type for all "colllection"-types?

**Comment by aartaka** [comment]
> * Looks like `lisp-url` should no longer be used.  Its only call site is `send-lisp-url`.  Move it there?
> * Actually `download.lisp` still uses `lisp-url` and as such all buttons are broken.

Yes and yes, just noticed it too!

**Comment by aartaka** [comment]
> 3\. If I got it right, this means that `user-interface:button` can be removed.

Why so?

**Comment by aartaka** [comment]
Would you be interested in fixing those, Pierre, or should I try it tomorrow?

**Comment by Ambrevar** [comment]
> Hash-tables are to be checked too. Actually, is there an umbrella type for all "colllection"-types?

Not that I know of.

> > 3\. If I got it right, this means that `user-interface:button` can be removed.

> Why so?

Because it's not used anywhere, it seems ;)

> Would you be interested in fixing those, Pierre, or should I try it tomorrow?

I won't have time today.  I can fix it tomorrow if you prefer, as you want.

**Comment by Ambrevar** [comment]
About the hash-table: it's a structure object, so the code I've proposed works.

**Comment by Ambrevar** [comment]
My bad, this is not portable, the specs say the class precedence is `hash-table -> T`.  It's not a structure object on CCL.

**Comment by aartaka** [comment]
2 (download.lisp) should be fixed with b7d33a559d8ef9b269ccf9a3d4d36cde8b027a11.

**Comment by aartaka** [comment]
`user-interface:button` is not to be removed -- it's still useful.
